Add proposal "Format API"

mickael-menu opened the pull request:
This proposal introduces a dedicated API to easily figure out a file format.

While a `Publication` is independent of any particular format, knowing the format of a publication file is necessary to:

* determine the publication parser to use,
* group or search publications by file type in the user's bookshelf.

This API is not tied to `Publication`, so it can be used as a general purpose tool to guess a file format, e.g. during HTTP requests or in the LCP library.

[You can read the formatted proposal here](https://github.com/mickael-menu/architecture/blob/proposal/file-format-api/proposals/001-format-api.md).

mickael-menu commented on proposals/001-file-format-api.md:
```diff
@@ -0,0 +1,608 @@
+
+Constant | Name | Extension | Media Type
+-------- | ---- | --------- | ----------
+`Audiobook` | Audiobook | audiobook | application/audiobook+zip
```
@llemeurfr You might want to review the name of each format, especially for the LCP and RWPM formats.

mickael-menu commented on proposals/001-file-format-api.md:
```diff
@@ -0,0 +1,608 @@
+`Pdf` | PDF | pdf | application/pdf
+`WebPub` | Web Publication | webpub | application/webpub+zip
+`WebPubManifest` | Web Publication | json | application/webpub+json
+`Zab` | Zipped Audio Book | zab | application/vnd.zab+zip
```
I couldn't find any media type online for ZAB, so I created this one from scratch: `application/vnd.zab+zip`

mickael-menu replied:
I used the `x.` facet instead.

qnga commented on proposals/001-file-format-api.md:
```diff
@@ -0,0 +1,608 @@
+* `isManifest: Bool`
+  * Returns whether this media type is contained by `AudiobookManifest`, `DiViNaManifest`, `Opds1Entry`, `Opds2Publication` or `WebPubManifest`.
+  * Used to determine if the remote URL needs to be stored in the database, instead of the local downloaded path.
+* `isOpds: Bool`
```
Maybe a more generic name could be used: `Catalog` for example. OPDS is just one format.

mickael-menu replied:
Actually the use case is that a reading app will want to present a catalog with different views, depending on whether it's an HTML page or an OPDS catalog. So we need a way to differentiate them.

Hence, the "used to determine the type of remote catalogs" for both `isHtml` and `isOpds`.

qnga replied:
Hum, ok.

qnga commented on proposals/001-file-format-api.md:
```diff
@@ -0,0 +1,608 @@
+  * Used to determine the type of remote catalogs.
+* `isHtml: Bool`
+  * Returns whether this media type is contained by `Html` or `Xhtml`.
+  * Used to determine the type of remote catalogs.
```
Is this a typo?

mickael-menu replied:
See this comment: https://github.com/readium/architecture/pull/127#discussion_r397016376

mickael-menu commented:
@danielweck @jccr We would like to merge this proposal during the next Readium dev call, so any insights you may have for desktop and web is most welcomed 🙏

mickael-menu commented on proposals/001-format-api.md:
```diff
@@ -0,0 +1,629 @@
+* Light:
+  * media type is `application/opds-publication+json`
+* Heavy:
+  * it's a JSON file parsed as an RWPM with at least one `Link` with a rel *starting with* `http://opds-spec.org/acquisition` 
```
@HadrienGardeur I'm actually not sure that an OPDS 2 Publication *must* have a an acquisition link. For example, for an audiobook it could be only that, right?
```json
"links": [
    {"rel": "self", "href": "http://example.org/manifest.json", "type": "application/audiobook+json"}
]
```

In which case, I guess there's no way to differentiate an OPDS 2 Publication from another RWPM, except from the context?

HadrienGardeur replied:
An OPDS 2.0 publication must have either:

- an acquisition link
- or a `readingOrder`

mickael-menu replied:
Okay, then we can keep this "heavy" step, but if we don't have the media type or an acquisition link, then the sniffing will return `WebPub` instead of `OPDS2Publication`.

mickael-menu commented:
I'd like to move forward, so I intend to merge this proposal after next week's meeting, if there's no counter-arguments till then.

I made a few changes in the proposal after [implementing it in Swift](https://github.com/readium/r2-shared-swift/pull/88):

* Renamed `Format.guess` into `Format.of`, because the caller is expecting an accurate `Format` returned.
* Added bitmap formats sniffing, because we need them in the CBZ parser.
* Removed `inspectingContent` parameter to simplify the sniffers. Instead, `Format.of()` will iterate twice through the sniffers: first with a context containing only file extensions and media types, and the second time with a context containing the content, if there's any.
* I grouped the sniffing of several formats in single sniffers, when the logic is shared (e.g. all the Readium WebPub formats).
* Added a few more APIs:
  * `MediaType`: `encoding`, `isRWPM`, `isPartOf()`
  * `Link`: `mediaType`,
  * `Format.SnifferContext`: `encoding`,`contentAsRWPM` and `readFileSignature()` to sniff magic numbers.
